docs/proposals/vip-4.md

```diff
@@ -0,0 +1,217 @@
+reprocessing "for free" as well.
+
+The goal is for lifecycle hooks to achieve the following use cases:
+
+* Orchestrate how data is served in each region, including:
+  * Ensuring a minimum delay (e.g. 1 hour) between each region beginning to serve the new store-version.
+  * Delaying the swapping of a new store-version to be within some time of day (e.g. during "business hours").
+  * Performing custom health checks on the client applications to ensure that their key metrics are still healthy within
+    a region where a new store-version was swapped, before proceeding to more regions. Based on the outcome of this 
+    check:
+    * Having the ability to abort the swapping of a store-version to further regions.
+    * Having the ability to rollback to the previous store-version in regions that already swapped.
+* Trigger informational events in proprietary monitoring systems after important lifecycle milestones are completed.
+
+## Project Justification
+
+The cost of having global impact in the case of issues mentioned above is too high, and we would like to provide 
+first-class options to reduce the blast radius. Building this within Venice itself will make it easier to automate these
+methodologies, thus reducing toil for users.
+
+## Functional specification
+
+The proposed API for this functionality is described in code here:
+
+[//]: # (Got to remove the /venice prefix before pushing to main on linkedin/venice, otherwise links will be broken...)
+
+* [StoreLifecycleHooks](/venice/javadoc/com/linkedin/venice/hooks/StoreLifecycleHooks.html), which is the main part of this 
+  proposal.
+* [StoreLifecycleEventOutcome](/venice/javadoc/com/linkedin/venice/hooks/StoreLifecycleEventOutcome.html), which is the signal
+  returned by some hooks to indicate that a given step should proceed or abort.
+* [StoreVersionLifecycleEventOutcome](/venice/javadoc/com/linkedin/venice/hooks/StoreVersionLifecycleEventOutcome.html), which
+  is the signal returned by some other hooks which need more fine-grained control over the workflow. In addition to 
+  proceeding and aborting, this also provides the option to wait, which tells the hooks framework to try invoking the 
+  hook again later, and rollback, which tells the framework to rollback to the previous store-version in all regions.
+
+## Proposed Design
+
+The main design consideration is where to execute the hooks. At a high level, there are three options:
+
+1. Within the push job.
+2. Within the parent controller.
+3. Within the child controllers.
+
+It is also possible to consider invoking some hooks in one of these location while other hooks would be executed 
+elsewhere. The table below summarizes the feasibility and tradeoffs for each of the proposed hooks:
+
+|                                              | Child Controller | Parent Controller | Push Job |
+|---------------------------------------------:|:----------------:|:-----------------:|:--------:|
```

Comment:
can we consider safety of running the validation hook in various components ? vpj would be the most safe ?

Reply:
Yes, that is true.

For hooks needing complex logic, we may want to isolate their business logic into a separate service, and have the Venice-side hook be just a thin client for interacting with the other service...

But if we were to pack a lot of logic into the hook itself, then the more widely distributed and per-user nature of VPJ would indeed be safer...

But running inside VPJ for fine-grained control of push behaviors is also potentially tedious... it seems like it might be easier to build the hooks framework in the location which already contains a lot of the relevant logic and state. But that's not an obvious conclusion, and certainly, all options are feasible.

Reply:
I guess the solution can be a combination of collaboration among several components.
For example, VPJ can be the orchestrator to maintain the state of the lifecycle execution, and Controller can be a stateless component for lifecycle management, but it just responds to the VPJ ask to execute the corresponding hooks/runtime blindly if required (or we think some hooks are better or more convenient to be executed in Controller instead of VPJ) without maintaining the stage of the lifecycle management.

Reply:
There is a benefit to let Parent Controller to be the orchestrator, stream processing job can be easily supported since it doesn't have an executor of push job as VPJ.

Reply:
> I guess the solution can be a combination of collaboration among several components.

Yep, it can be.

> There is a benefit to let Parent Controller to be the orchestrator, stream processing job can be easily supported since it doesn't have an executor of push job as VPJ.

That's right... and there are other similar issues related to using VPJ instead of the parent controller.

For example, schema registration (which we would like to emit a monitoring event for) would be caught in all cases by the controller but not by VPJ (there are other solutions of course, such as emitting the monitoring even in the other components besides VPJ that can register schemas, but it's potentially whack-a-mole, whereas the controller will for sure see all schema registration calls).

Another example is version swap... it is possible to invoke this via the controller client / admin tool outside of a push job, and we may want to invoke hooks in those cases too (at least for monitoring event purposes, but possibly for flow control purposes, although for flow control we might want to support an override to "skip hooks" or override specific hook configs, when calling from the admin tool...).

Reply:
Thats a great point, given users planning to use business hour version swap thru controller client (probably thats not part of the scope here), having hooks to validate the action would be great. Though we need to have a way to communicate/rollback the action to the user.